File: src/018FourSum/018FourSum.go
package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int, t int) [][]int {
	var r [][]int
	n := len(nums)
	for k := 0; k < n-2; k++ {
		i := k + 1
		j := n - 1
		target := t - nums[k]
		for i < j {
			if nums[i]+nums[j] == target {
				r = append(r, []int{nums[k], nums[i], nums[j]})
				for i < j && nums[i] == nums[i+1] {
					i++
				}
				i++
				for i < j && nums[j-1] == nums[j] {
					j--
				}
				j--
			} else if nums[i]+nums[j] < target {
				for i < j && nums[i] == nums[i+1] {
					i++
				}
				i++
			} else {
				for i < j && nums[j-1] == nums[j] {
					j--
				}
				j--
			}
		}
		for k < n-2 && nums[k] == nums[k+1] {
			k++
		}
	}
	return r
}

func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	var r4 [][]int
	n := len(nums)
	for k := n - 1; k >= 3; k-- {
		r3 := threeSum(nums[:k], target-nums[k])
		if r3 != nil {
			for i := 0; i < len(r3); i++ {
				r4 = append(r4, []int{r3[i][0], r3[i][1], r3[i][2], nums[k]})
			}
		}
		for k >= 3 && nums[k] == nums[k-1] {
			k--
		}
	}
	return r4
}

func main() {
	fmt.Printf("%v\n", fourSum([]int{1, 0, -1, 0, -2, 2}, 0))
}

File: src/018FourSum/018FourSum.rs
impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut nums = nums;
        let mut result = vec![];
        let mut i = 0;
        
        nums.sort();
        
        while i < nums.len() as i32 - 3 {
            let mut j = i+1;
            while j < nums.len() as i32 - 2 {
                let (mut k, mut l) = (j+1, nums.len() as i32 - 1);
                while k < l {
                    let sum = nums[i as usize] + nums[j as usize] + nums[k as usize] + nums[l as usize];
                    if sum == target {
                        result.push(vec![nums[i as usize], nums[j as usize], nums[k as usize], nums[l as usize]]);
                        k += 1;
                        while k < l && nums[k as usize] == nums[k as usize-1] {
                            k +=1;
                        }
                    }else if sum < target {
                        k += 1;
                    }else{
                        l -= 1;
                    }
                }
                j += 1;
                while j < nums.len() as i32 - 2 && nums[j as usize] == nums[j as usize-1] {
                    j += 1;
                }
            }
            
            i += 1;
            while i < nums.len() as i32 - 3 && nums[i as usize] == nums[i as usize-1] {
                i += 1;
            }
        }
        return result;
    }
}
